lib/sqlite/sqlite.go

```diff
@@ -0,0 +1,165 @@
+
+var pragmas = []string{
+	"PRAGMA synchronous = normal",
+	"PRAGMA temp_store = memory",
+	"PRAGMA mmap_size = 30000000000",
+	"PRAGMA page_size = 32768",
+	"PRAGMA auto_vacuum = NONE",
+	"PRAGMA automatic_index = OFF",
+	"PRAGMA journal_mode = WAL",
+	"PRAGMA wal_autocheckpoint = 256", // checkpoint @ 256 pages
+	"PRAGMA journal_size_limit = 0",   // always reset journal and wal files
+}
+
+const metaTableDdl = `CREATE TABLE IF NOT EXISTS _meta (
+	version UINT64 NOT NULL UNIQUE
+)`
+
+// metaDdl returns the DDL statements required to create the _meta table and add the required
+// up to the given version.
+func metaDdl(version uint64) []string {
+	var ddls []string
+	for i := 1; i <= int(version); i++ {
+		ddls = append(ddls, `INSERT OR IGNORE INTO _meta (version) VALUES (`+strconv.Itoa(i)+`)`)
+	}
+	return append([]string{metaTableDdl}, ddls...)
+}
+
+// Open opens a database at the given path. If the database does not exist, it will be created.
+func Open(path string) (*sql.DB, bool, error) {
+	if err := os.MkdirAll(filepath.Dir(path), 0755); err != nil {
+		return nil, false, xerrors.Errorf("error creating database base directory [@ %s]: %w", path, err)
+	}
+
+	_, err := os.Stat(path)
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return nil, false, xerrors.Errorf("error checking file status for database [@ %s]: %w", path, err)
+	}
+	exists := err == nil
+
+	db, err := sql.Open("sqlite3", path+"?mode=rwc")
+	if err != nil {
+		return nil, false, xerrors.Errorf("error opening database [@ %s]: %w", path, err)
+	}
+
+	for _, pragma := range pragmas {
+		if _, err := db.Exec(pragma); err != nil {
+			_ = db.Close()
+			return nil, false, xerrors.Errorf("error setting database pragma %q: %w", pragma, err)
+		}
+	}
+
+	return db, exists, nil
+}
+
+// InitDb initializes the database by checking whether it needs to be created or upgraded.
+// The ddls are the DDL statements to create the tables in the database and their initial required
+// content. The schemaVersion will be set inside the databse if it is newly created. Otherwise, the
+// version is read from the databse and returned. This value should be checked against the expected
+// version to determine if the database needs to be upgraded.
+func InitDb(
+	ctx context.Context,
+	name string,
+	db *sql.DB,
+	ddls []string,
+	versionMigrations []MigrationFunc,
+) error {
+
+	schemaVersion := len(versionMigrations) + 1
+
+	q, err := db.QueryContext(ctx, "SELECT name FROM sqlite_master WHERE type='table' AND name='_meta';")
+	if q != nil {
+		defer func() { _ = q.Close() }()
+	}
+
+	if errors.Is(err, sql.ErrNoRows) || !q.Next() {
+		// empty database, create the schema including the _meta table and its versions
+		ddls := append(metaDdl(uint64(schemaVersion)), ddls...)
+		for _, ddl := range ddls {
+			if _, err := db.Exec(ddl); err != nil {
+				return xerrors.Errorf("failed to %s database execute ddl %q: %w", name, ddl, err)
```

Comment:
I think we also need a `db.Close` here

```
_ = db.Close()
```

Reply:
@rvagg Do we need to close the DB here before returning ?

Reply:
should be done in the calling function; it was in 2 of the 4 calls, but I've added the other 2 and a comment in the method docs
good catch!